catalog/imaging/imaging-postprocess-plugin-nitf/src/main/java/org/codice/alliance/plugin/nitf/NitfPostProcessPlugin.java

```diff
@@ -216,49 +233,86 @@ private void handleProcessCreateItem(List<ProcessCreateItem> processCreateItems)
+                  processResource,
+                  constructorTriple ->
+                      new ProcessCreateItemImpl(
+                          constructorTriple.getLeft(), constructorTriple.getMiddle()));
+        } finally {
+          lock.release();
         }
-      } catch (IOException | NitfFormatException | RuntimeException e) {
-        LOGGER.debug("An error occured when rendering a nitf for {}", processResource.getName(), e);
       } catch (InterruptedException e) {
-        LOGGER.error("Rendering failed for {}", processResource.getName(), e);
+        LOGGER.debug("Interrupt received while doing image processing.", e);
         Thread.currentThread().interrupt();
-        throw new RuntimeException(
-            String.format("Rendering failed for %s", processResource.getName()), e);
       }
     }
+
+    if (createdItems == null) {
+      createdItems = new ArrayList<>();
+    }
     return createdItems.stream();
   }
 
+  private <T extends ProcessResourceItem> List<T> process(
+      T processResourceItem,
+      Metacard metacard,
+      Metacard originalMetacard,
+      ProcessResource processResource,
+      Function<Triple<ProcessResource, Metacard, Metacard>, T> constructor) {
+    List<T> items = new ArrayList<>();
+    try (TemporaryFileBackedOutputStream fbos =
+        new TemporaryFileBackedOutputStream(DEFAULT_FILE_BACKED_OUTPUT_STREAM_THRESHOLD)) {
+
+      // TODO: 06/21/2018 oconnormi - can probably get rid of this once
+      // ProcessResourceImpl.getInputStream can be called multiple times
+      fbos.write(IOUtils.toByteArray(processResource.getInputStream()));
+      ByteSource byteSource = fbos.asByteSource();
+      BufferedImage renderedImage = renderImage(byteSource.openStream());
+
+      if (renderedImage != null) {
+        addThumbnailToMetacard(metacard, renderedImage);
+        processResourceItem.markMetacardAsModified();
+        if (createOverview) {
+          ProcessResource overviewProcessResource = createOverviewResource(renderedImage, metacard);
+          items.add(
+              constructor.apply(
+                  new ImmutableTriple<>(overviewProcessResource, metacard, originalMetacard)));
+        }
+
+        if (storeOriginalImage) {
+          ProcessResource originalImageProcessResource =
+              createOriginalImage(
+                  renderImageUsingOriginalDataModel(byteSource.openStream()), metacard);
+
+          items.add(
+              constructor.apply(
+                  new ImmutableTriple<>(originalImageProcessResource, metacard, originalMetacard)));
+        }
+      }
+    } catch (IOException | NitfFormatException | RuntimeException e) {
+      LOGGER.debug("An error occured when rendering a nitf for {}", processResource.getName(), e);
+    } catch (InterruptedException e) {
+      LOGGER.error("Rendering failed for {}", processResource.getName(), e);
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(
```

Comment:
❓ Do we want to leak the stacktrace here, or just include the exception's message? It's already being logged to the server.